store/hbase/txn.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hbasekv
+
+import (
+	"fmt"
+	"runtime/debug"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/go-hbase"
+	"github.com/pingcap/go-themis"
+	"github.com/pingcap/tidb/kv"
+)
+
+var (
+	_ kv.Transaction = (*hbaseTxn)(nil)
+	// ErrInvalidTxn is the error when commits or rollbacks in an invalid transaction.
+	ErrInvalidTxn = errors.New("invalid transaction")
+	// ErrCannotSetNilValue is the error when sets an empty value
+	ErrCannotSetNilValue = errors.New("can not set nil value")
+)
+
+// dbTxn is not thread safe
+type hbaseTxn struct {
+	kv.UnionStore
+	*themis.Txn
+	store     *hbaseStore // for commit
+	storeName string
+	tid       uint64
+	valid     bool
+	version   kv.Version // commit version
+}
+
+func newHbaseTxn(t *themis.Txn, storeName string) *hbaseTxn {
+	return &hbaseTxn{
+		Txn:        t,
+		valid:      true,
+		storeName:  storeName,
+		tid:        t.GetStartTS(),
+		UnionStore: kv.NewUnionStore(newHbaseSnapshot(t, storeName)),
+	}
+}
+
+// Implement transaction interface
+
+func (txn *hbaseTxn) Inc(k kv.Key, step int64) (int64, error) {
+	log.Debugf("Inc %q, step %d txn:%d", k, step, txn.tid)
+	k = kv.EncodeKey(k)
+	val, err := txn.UnionStore.Get(k)
+	if kv.IsErrNotFound(err) {
+		err = txn.UnionStore.Set(k, []byte(strconv.FormatInt(step, 10)))
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		return step, nil
+	}
+
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	intVal, err := strconv.ParseInt(string(val), 10, 64)
+	if err != nil {
+		return intVal, errors.Trace(err)
+	}
+
+	intVal += step
+	err = txn.UnionStore.Set(k, []byte(strconv.FormatInt(intVal, 10)))
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	return intVal, nil
+}
+
+func (txn *hbaseTxn) String() string {
+	return fmt.Sprintf("%d", txn.tid)
+}
+
+func (txn *hbaseTxn) Get(k kv.Key) ([]byte, error) {
+	log.Debugf("get key:%q, txn:%d", k, txn.tid)
+	k = kv.EncodeKey(k)
+	val, err := txn.UnionStore.Get(k)
+	if kv.IsErrNotFound(err) || len(val) == 0 {
+		return nil, kv.ErrNotExist
```

Comment:
It is better to use `errors.Trace(kv.ErrNotExist)`, for many places return `kv.ErrNotExist`, so we should use juju/errors to trace error stack info.